release: add RedHat docker image

rail opened the pull request:
In the past we used a process which downloads a release tarball and then creates a new docker image from scratch. This process is manual and error prone.

This patch adds a new script which reuses the official docker image and adds some extra labels in order to match RedHat requirements.

* Add licenses to all docker images
* Add `version` label to all images

Release justification: non-production code changes

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60007)
<!-- Reviewable:end -->

rail commented:
@jlinder I would appreciate your feedback on this draft. I wonder if it aligns with your thoughts.

jlinder reviewed (commented):
Sure thing. Here's what I have.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-:-MT2VQ6O6qDQAZWhT2Oy:b-eovohv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rail)

___
*[build/deploy/Dockerfile, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2PwsC6HHLGBAWzAT9:-MT2PwsDDQRX_tUa8aVJ:bjd61d) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L31)):*
> ```
> LABEL summary="CockroachDB is a Distributed SQL database."
> LABEL description="CockroachDB is a PostgreSQL wire-compatable Distributed SQL database."
> ```

I realized these sentences probably weren't passed through docs before. Might be good to ask them what they would say here?
___
*[build/deploy/Dockerfile, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2OJTr6Nf7Wy9odUIj:-MT2OJTs2PgVKRsPwuU6:b-8wnsvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L32)):*
> ```
> # TODO(rail): add k8s/redshift specific labels too
> # io.k8s.display-name, io.k8s.description and io.openshift.tags
> # Action-oriented labels, such as run or stop.
> ```

What are these labels? Where do you see the need for them coming from?
___
*[build/deploy/Dockerfile, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Q4eL0nn0sSzX7s47:-MT2Q4eL0nn0sSzX7s48:bphgc05) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L38)):*
> ```Dockerfile
> EXPOSE 26257 8080
> # TODO(rail): verify this works as expected
> USER 1001:1001
> ```

Why add this?
___
*[build/release/teamcity-publish-release.sh, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Qs7xC4mW8bcGi3YH:-MT2Qs7xC4mW8bcGi3YI:b-yfgxeb) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L25)):*
> ```Shell
> release_branch=$(echo ${build_name} | grep -E -o '^v[0-9]+\.[0-9]+')
> version=$(echo ${build_name} | sed -e 's/^v//' | cut -d- -f 1)
> release=$(echo ${build_name} | sed -e 's/^v//' | cut -d- -f 2)
> ```

Release can always be 1.

From the certification requirements, release is "A number used to identify the specific build for this image". Following what Red Hat does for UBI docker images, that number is a strictly increasing number within a version as can be seen [in the image tags](https://catalog.redhat.com/software/containers/ubi8/ubi-minimal/5c359a62bed8bd75a2c3fba8). Since every one of our releases is a completely new version, we can just use 1 here (or perhaps 0).
___
*[build/release/teamcity-publish-release.sh, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2QCsHEta5SP9Sy0e3:-MT2QCsHEta5SP9Sy0e4:b8c0kz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L93)):*
> ```Shell
> cp cockroach lib/libgeos.so lib/libgeos_c.so build/deploy
> 
> # TODO(rail): do we need to copy the licenses?
> ```

The licenses need to be copied into the image for Red Hat. It's item 9 in the Red Hat certification requirements doc.
___
*[build/release/teamcity-publish-release.sh, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2ULdeBL1YOVfhojBK:-MT2ULdfD3H3c_GlhiZW:bdmmv37) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L125)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
>   # TODO(rail): create a dry-run env for rhel
>   # as an alternative these steps can be extracted into a separate process, so we
>   # don't fail the whole thing in case of vendor specific failures
>   # make sure that using our FROM: is acceptable by rhel
>   sed -e "s/@build_name@/${build_name}/g" build/redhat/Dockerfile.in > build/redhat/Dockerfile
> ```
</details>


Let's move this to a separate file and separate build configuration in TeamCity. I already want to split this file into multiple build configuration steps for that exact failure reason. There've been multiple times when a step later in this file (like pushing docker images) failed and all the binaries that didn't need to be rebuilt were force rebuilt on the subsequent run(s). Also, there are lots of other things that need to be automated and we want to make it easy for other teams to add steps to the release process.


<!-- Sent from Reviewable.io -->

rail reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/deploy/Dockerfile, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2PwsC6HHLGBAWzAT9:-MT2dvCGB_moG9Kw7_eR:bebrt1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L31)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

> ```
> LABEL summary="CockroachDB is a Distributed SQL database."
> LABEL description="CockroachDB is a PostgreSQL wire-compatable Distributed SQL database."
> ```

I realized these sentences probably weren't passed through docs before. Might be good to ask them what they would say here?
</blockquote></details>

Will do
___
*[build/deploy/Dockerfile, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2OJTr6Nf7Wy9odUIj:-MT2e3MxAGoYZiIJ63Np:b-gixhql) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L32)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

> ```
> # TODO(rail): add k8s/redshift specific labels too
> # io.k8s.display-name, io.k8s.description and io.openshift.tags
> # Action-oriented labels, such as run or stop.
> ```

What are these labels? Where do you see the need for them coming from?
</blockquote></details>

They were mentioned in the RedHat doc as recommended labels.
___
*[build/deploy/Dockerfile, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Q4eL0nn0sSzX7s47:-MT2eC7t7Q9mbQDXFC5v:b7252hv) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L38)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Why add this?
</blockquote></details>

To make sure  we don't run the main process as root. This is not a hard requirement, but a good-to-have security hygiene  feature. We will need to to verify this very carefully and make sure that crdb doesn't rely on root privileges.
___
*[build/release/teamcity-publish-release.sh, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Qs7xC4mW8bcGi3YH:-MT2emHNDpqnxMY-ZgeG:bymu847) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L25)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Release can always be 1.

From the certification requirements, release is "A number used to identify the specific build for this image". Following what Red Hat does for UBI docker images, that number is a strictly increasing number within a version as can be seen [in the image tags](https://catalog.redhat.com/software/containers/ubi8/ubi-minimal/5c359a62bed8bd75a2c3fba8). Since every one of our releases is a completely new version, we can just use 1 here (or perhaps 0).
</blockquote></details>

Hmmm, I wonder if we should allow passing in the release number with default set to 0 or 1 or just hardcode it to 0/1.
___
*[build/release/teamcity-publish-release.sh, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2QCsHEta5SP9Sy0e3:-MT2fEE4AOfd-z-fBGHu:b-kbp5k6) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L93)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

The licenses need to be copied into the image for Red Hat. It's item 9 in the Red Hat certification requirements doc.
</blockquote></details>

Sorry. The TODO isn't verbose enough. I wonder if we should include the licenses into *all* images, not only the redhat one.
___
*[build/release/teamcity-publish-release.sh, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2ULdeBL1YOVfhojBK:-MT2fS1_2Ousa-u0ANI-:b-5fzrj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L125)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

> ```
>   # TODO(rail): create a dry-run env for rhel
>   # as an alternative these steps can be extracted into a separate process, so we
>   # don't fail the whole thing in case of vendor specific failures
>   # make sure that using our FROM: is acceptable by rhel
>   sed -e "s/@build_name@/${build_name}/g" build/redhat/Dockerfile.in > build/redhat/Dockerfile
> ```

Let's move this to a separate file and separate build configuration in TeamCity. I already want to split this file into multiple build configuration steps for that exact failure reason. There've been multiple times when a step later in this file (like pushing docker images) failed and all the binaries that didn't need to be rebuilt were force rebuilt on the subsequent run(s). Also, there are lots of other things that need to be automated and we want to make it easy for other teams to add steps to the release process.
</blockquote></details>

SGTM! I'll split them.


<!-- Sent from Reviewable.io -->

jlinder reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/deploy/Dockerfile, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2OJTr6Nf7Wy9odUIj:-MT3IKfZEFYyrx1jH5kV:b4lf6dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L32)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

They were mentioned in the RedHat doc as recommended labels.
</blockquote></details>

Ah. I see them now. If you end up researching and adding them, perhaps we do it in the `build/redhat/Dockerfile.in` file to start?
___
*[build/deploy/Dockerfile, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Q4eL0nn0sSzX7s47:-MT3J39xB4Tso7GvI84y:b-ag59i6) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L38)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

To make sure  we don't run the main process as root. This is not a hard requirement, but a good-to-have security hygiene  feature. We will need to to verify this very carefully and make sure that crdb doesn't rely on root privileges.
</blockquote></details>

We've previously had discussion about to root or not to root. There are some questions that need answering to not root. See [this slack conversation](https://cockroachlabs.slack.com/archives/CGA9F858R/p1602105807086600).
___
*[build/redhat/Dockerfile.in, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT3Irrx3vy8gfIwR_W6:-MT3Irrx3vy8gfIwR_W7:bacy7do) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/redhat/Dockerfile.in#L1)):*
> ```
> FROM cockroachdb/cockroach:@build_name@
> ```

Perhaps we put this file in the `build/deploy-redhat` directory? Namespace it to be by deploy and visibly linked to deploy.
___
*[build/release/teamcity-publish-release.sh, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Qs7xC4mW8bcGi3YH:-MT3JLKT6DLRzeXXjsYO:bjq563z) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L25)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Hmmm, I wonder if we should allow passing in the release number with default set to 0 or 1 or just hardcode it to 0/1.
</blockquote></details>

I don't see it being used any time soon for a couple reasons. 

1) Using it would require us to opt-in to a versioning system similar to how RedHat does it. 
2) We release a given version only once. If any changes are required, that becomes a new version.
3) We use a versioning scheme that has everything in it we need to capture the version. It's semver with major = the year of the release series, minor = the release series in the year (1 in spring, 2 in fall), and patch as being the iteration.

Also, I'd like for the parameter set offered to the release manager by the build configuration to be only what is necessary to do the release types we need. A couple of our goals are to minimize the possibility for human errors in the release process and to make the release process so simple that a non-engineer can make a release (though that is quite a ways off). 

If we end up changing the versioning scheme and this becomes relevant, then we can add it in then.
___
*[build/release/teamcity-publish-release.sh, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2QCsHEta5SP9Sy0e3:-MT3NwKoEJ4C4PvJfw1N:b-4e667h) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L93)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Sorry. The TODO isn't verbose enough. I wonder if we should include the licenses into *all* images, not only the redhat one.
</blockquote></details>

Ah. I see. I don't think we _need_ to put them in all images though I don't think it's a bad idea. When I proposed the idea to make the main dockerhub image contain all the stuff needed for the Red Hat image, I was thinking to put the licenses in the initial image.


<!-- Sent from Reviewable.io -->

rail reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/deploy/Dockerfile, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2OJTr6Nf7Wy9odUIj:-MT3Wbf4EJppIkJ7bBss:b-ok4mo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L32)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Ah. I see them now. If you end up researching and adding them, perhaps we do it in the `build/redhat/Dockerfile.in` file to start?
</blockquote></details>

Sure. Do you think I should also move the release and version labels there too?
___
*[build/deploy/Dockerfile, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Q4eL0nn0sSzX7s47:-MT3WoiQ2hhC6txQ3fKF:b-5yldx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L38)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

We've previously had discussion about to root or not to root. There are some questions that need answering to not root. See [this slack conversation](https://cockroachlabs.slack.com/archives/CGA9F858R/p1602105807086600).
</blockquote></details>

Ah, I see the point in that conversation - --user is your friend if you want to run it as a user. Right now RedHat only recommends this approach, but they may enforce this in the future (per doc). Probably it's worth filing a separate issue about using USER in Dockerfile. I'll remove that line for now.
___
*[build/redhat/Dockerfile.in, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT3Irrx3vy8gfIwR_W6:-MT3XWD1C-k1B67X1ICm:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/redhat/Dockerfile.in#L1)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Perhaps we put this file in the `build/deploy-redhat` directory? Namespace it to be by deploy and visibly linked to deploy.
</blockquote></details>

:thumbsup: 
___
*[build/release/teamcity-publish-release.sh, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Qs7xC4mW8bcGi3YH:-MT3XlbgBzynzdKsvyoK:bip74av) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L25)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I don't see it being used any time soon for a couple reasons. 

1) Using it would require us to opt-in to a versioning system similar to how RedHat does it. 
2) We release a given version only once. If any changes are required, that becomes a new version.
3) We use a versioning scheme that has everything in it we need to capture the version. It's semver with major = the year of the release series, minor = the release series in the year (1 in spring, 2 in fall), and patch as being the iteration.

Also, I'd like for the parameter set offered to the release manager by the build configuration to be only what is necessary to do the release types we need. A couple of our goals are to minimize the possibility for human errors in the release process and to make the release process so simple that a non-engineer can make a release (though that is quite a ways off). 

If we end up changing the versioning scheme and this becomes relevant, then we can add it in then.

</blockquote></details>

It makes sense. My only concern left is what happens if we need to rebuild the same version due to some security update in the base image? As a workaround we can bump the release digit, but that would require a new branch.
___
*[build/release/teamcity-publish-release.sh, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2QCsHEta5SP9Sy0e3:-MT3YtQ80K0Yzym8tkGl:b38fi6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L93)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Ah. I see. I don't think we _need_ to put them in all images though I don't think it's a bad idea. When I proposed the idea to make the main dockerhub image contain all the stuff needed for the Red Hat image, I was thinking to put the licenses in the initial image.
</blockquote></details>

Copy that. :thumbsup: 


<!-- Sent from Reviewable.io -->

rail committed:
release: automate publishing RedHat images

In the past we used a process which downloads a release tarball
and then creates a new docker image from scratch. This process is
manual and error prone.

This patch adds a new script which reuses the official docker image and
adds some extra labels in order to match RedHat requirements.

* Add licenses to all docker images
* Add `version` label to all images

Release justification: non-production code changes

Release note: None

rail reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/deploy/Dockerfile, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Q4eL0nn0sSzX7s47:-MT6xQ-h7TTS6lq4dHbL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L38)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Ah, I see the point in that conversation - --user is your friend if you want to run it as a user. Right now RedHat only recommends this approach, but they may enforce this in the future (per doc). Probably it's worth filing a separate issue about using USER in Dockerfile. I'll remove that line for now.
</blockquote></details>

Done.
___
*[build/redhat/Dockerfile.in, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT3Irrx3vy8gfIwR_W6:-MT6xUKB-S9vczFWRwXn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/redhat/Dockerfile.in#L1)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

:thumbsup: 
</blockquote></details>

Done.
___
*[build/release/teamcity-publish-release.sh, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Qs7xC4mW8bcGi3YH:-MT6xWRE-DCMI3G4q0de:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L25)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

It makes sense. My only concern left is what happens if we need to rebuild the same version due to some security update in the base image? As a workaround we can bump the release digit, but that would require a new branch.
</blockquote></details>

Done.
___
*[build/release/teamcity-publish-release.sh, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2QCsHEta5SP9Sy0e3:-MT6xY183Z_0E3CiZ3uF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L93)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Copy that. :thumbsup: 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rail commented:
I think I'm close to start testing this code. \o/

jlinder reviewed (commented):
While the changes now in this PR are manageable as code and something like this is where we'd like to get to, successfully making them live in TeamCity all at once is a big challenge. The reasons coming to mind are:

1. We need to backport these changes to the `release-20.1` and `release-20.2` branches. No changes would be needed to backport to 20.2. But 20.1 would need a bunch of changes (it uses Ubuntu Xenial for it's base docker image and it doesn't have the libgeos* files).

2. We _could_ call all the scripts sequentially from the current build configuration, but ideally we'd want to split them out into multiple build configurations including a new entry build configuration (i.e., the build configuration that the release engineer uses to trigger the release). 

That's a lot to build and test all at once and to have land at the same time (or to maintain two separate processes depending on the branch being released). This is a lot more work than the original issue was scoped for. And I'd like to communicate in writing what the new structure will be to other folks that care about it in the org before just doing it (it's a critical system; the structure is something people from other teams will be adding new steps into).

I'd recommend making smaller steps to get to the end goal.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-:-MThiYF_EpdZKDnceNsI:brmq43u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/deploy/Dockerfile, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2OJTr6Nf7Wy9odUIj:-MT7uUdS5V17zV9gaFIW:bwlk1zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L32)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Sure. Do you think I should also move the release and version labels there too?
</blockquote></details>

I'd move the release to `build/redhat/Dockerfile.in` and leave version.
___
*[build/release/release-support.sh, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThoSo0-WCCQRR6LV9p:-MThoSo1BvpSZ1eJLtgW:b-gv612k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/release-support.sh#L1)):*
> ```Shell
> # Common helpers for release-*.sh scripts.
> ```

I think we should not centralize all the variable setup and only centralize what is used in multiple files. The reasons:

- There's a lot in this file that is used for only one of the resulting scripts.
- There will also be a lot more scripts that people from other teams will be writing based on these examples. If we just throw it all together in one support script, this file will become a mess quickly.

If there are things that we want centralized, let's put them in functions instead so a given script can choose what it needs. 

Also, I'd recommend following naming pattern for scripts used in teamcity to start with `teamcity-` and rename this to `build/release/teamcity-release-support.sh` (assuming there is appropriate information to centralize).
___
*[build/release/teamcity-publish-redhat-release.sh, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThuLBMBDcVEOz8GjeK:-MThuLBN-53s4uJjtKvD:bzh2ln2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-publish-redhat-release.sh#L11)):*
> ```Shell
> source "$(dirname "${0}")/release-support.sh"
> tc_end_block "Variable Setup"
> ```

This script should also not allow pushing of pre-release versions to Red Hat.
___
*[build/release/teamcity-publish-redhat-release.sh, line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThsgLrBbWaMjKOm7V-:-MThsgLrBbWaMjKOm7V0:bnggzb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-publish-redhat-release.sh#L14)):*
> ```Shell
> tc_start_block "Configure docker"
> configure_docker_creds
> docker_login_with_google
> ```

Why does gcr need to be logged into here? Nothing is being pushed to or pulled from there.
___
*[build/release/teamcity-publish-redhat-release.sh, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThtlzn9MrCAtMQLFdv:-MThtlzn9MrCAtMQLFdw:bz6dcgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-publish-redhat-release.sh#L30)):*
> ```Shell
>   # TODO(rail): docker push
> else
>   # TODO(rail): create a dry-run env for rhel
> ```

I think it should just push to red hat but make the tag `${rhel_repository}:${build_name}-dryrun` to indicate it was from a dry run and should not be published.
___
*[build/release/teamcity-publish-redhat-release.sh, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MTht1cHBoF9y2sFatWs:-MTht1cHBoF9y2sFatWt:b-o9zqhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-publish-redhat-release.sh#L36)):*
> ```Shell
> tc_end_block "Rebuild docker image"
> 
> tc_start_block "Verify docker image"
> ```

The `tc_end_block` is missing for this section.
___
*[build/release/teamcity-publish-release.sh, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2Qs7xC4mW8bcGi3YH:-MThsDGS0hHGB2kgrJ6I:b9zmei4) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L25)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The security update would be a completely new cockroach version that would go through the normal release process.
___
*[build/release/teamcity-support.sh, line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThrKop7jhzXx1Nz58S:-MThrKop7jhzXx1Nz58T:bq4qzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-support.sh#L6)):*
> ```Shell
>   rm -f .google-credentials.json
>   rm -f .cockroach-teamcity-key
>   rm -rf ~/.docker
> ```

Can we do this in a separate PR? It's a good thing to get out quickly and will be easier to test + backport in standalone.


<!-- Sent from Reviewable.io -->

rail reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/release/teamcity-support.sh, line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThrKop7jhzXx1Nz58S:-MUOx5ii74RbGwHzK5Ob:b-yfdbpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-support.sh#L6)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Can we do this in a separate PR? It's a good thing to get out quickly and will be easier to test + backport in standalone.
</blockquote></details>

done: https://github.com/cockroachdb/cockroach/pull/61122


<!-- Sent from Reviewable.io -->

rail reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/deploy/Dockerfile, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2OJTr6Nf7Wy9odUIj:-MUiEeYjE7TH2nQr2koc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/deploy/Dockerfile#L32)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I'd move the release to `build/redhat/Dockerfile.in` and leave version.
</blockquote></details>

Done.
___
*[build/release/release-support.sh, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThoSo0-WCCQRR6LV9p:-MUiEn5o0r7C7VhkrWnz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/release-support.sh#L1)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I think we should not centralize all the variable setup and only centralize what is used in multiple files. The reasons:

- There's a lot in this file that is used for only one of the resulting scripts.
- There will also be a lot more scripts that people from other teams will be writing based on these examples. If we just throw it all together in one support script, this file will become a mess quickly.

If there are things that we want centralized, let's put them in functions instead so a given script can choose what it needs. 

Also, I'd recommend following naming pattern for scripts used in teamcity to start with `teamcity-` and rename this to `build/release/teamcity-release-support.sh` (assuming there is appropriate information to centralize).

</blockquote></details>

Done.
___
*[build/release/teamcity-publish-redhat-release.sh, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThuLBMBDcVEOz8GjeK:-MUiEuJl-tJpxHS9FhTs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-publish-redhat-release.sh#L11)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

This script should also not allow pushing of pre-release versions to Red Hat.
</blockquote></details>

Done.
___
*[build/release/teamcity-publish-redhat-release.sh, line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThsgLrBbWaMjKOm7V-:-MUiEvIr8LuNgx-z9I0f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-publish-redhat-release.sh#L14)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Why does gcr need to be logged into here? Nothing is being pushed to or pulled from there.
</blockquote></details>

Done.
___
*[build/release/teamcity-publish-redhat-release.sh, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MThtlzn9MrCAtMQLFdv:-MUiEwnz-KfJKnV6tRDT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-publish-redhat-release.sh#L30)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I think it should just push to red hat but make the tag `${rhel_repository}:${build_name}-dryrun` to indicate it was from a dry run and should not be published.
</blockquote></details>

Done.
___
*[build/release/teamcity-publish-redhat-release.sh, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MTht1cHBoF9y2sFatWs:-MUiExm35z6MbwongETK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3e5948340d680502b978dd0ad2f41080506e813/build/release/teamcity-publish-redhat-release.sh#L36)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

The `tc_end_block` is missing for this section.
</blockquote></details>

Done.
___
*[build/release/teamcity-publish-release.sh, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MT2ULdeBL1YOVfhojBK:-MUiF9qI2mBETOd5VJiM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42da3e4e053f5e425cf3a0566ef07479a5eb55bb/build/release/teamcity-publish-release.sh#L125)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

SGTM! I'll split them.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rail reviewed (commented):
I think this is ready for the final review.
I tested this as much I could in the "dev" environment.

I set up a new build configuration with all needed parameters and tested things by hardcoding some variables in the script in order to test a particular behaviour, that is hard to test by passing parameters. For example, I didn't want to run a dry-run using the main publishing script in order to create and push a docker image to the "misc" dockerhub repo and a bunch of other places. Instead I carefully changed the logic to use the published release docker image.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-:-MUjcQr33KATQdCEomT4:b-qr729h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)



<!-- Sent from Reviewable.io -->

jlinder reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rail)

___
*[build/deploy-redhat/Dockerfile.in, line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MUu77-SBkjyYozSUEdk:-MUu77-SBkjyYozSUEdl:bzfv1v3) ([raw file](https://github.com/cockroachdb/cockroach/blob/79cad1aeea145c3762ea28ae729395f0b876d88a/build/deploy-redhat/Dockerfile.in#L6)):*
> ```
> LABEL summary="CockroachDB is a Distributed SQL database."
> LABEL description="CockroachDB is a PostgreSQL wire-compatable Distributed SQL database."
> ```
Were you able to ask docs if these were good strings to use?
___
*[build/release/teamcity-publish-redhat-release.sh, line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MUu0XvX9TzJ7HeZeG-s:-MUu0XvX9TzJ7HeZeG-t:bbjdpnz) ([raw file](https://github.com/cockroachdb/cockroach/blob/79cad1aeea145c3762ea28ae729395f0b876d88a/build/release/teamcity-publish-redhat-release.sh#L10)):*
> ```Shell
> tc_start_block "Sanity Check"
> # Make sure that the version is a substring of TC branch name (e.g. v20.2.4-57-abcd345)
> if [[ $TC_BUILD_BRANCH != "$NAME*" ]]; then
> ```

I don't think this does what the comment and message say this intends to do. Testing locally, this doesn't give the results expected here.
___
*[build/release/teamcity-publish-redhat-release.sh, line 15 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MUu2r61A7qJduPESmp9:-MUu2r61A7qJduPESmpA:b-cp13q0) ([raw file](https://github.com/cockroachdb/cockroach/blob/79cad1aeea145c3762ea28ae729395f0b876d88a/build/release/teamcity-publish-redhat-release.sh#L15)):*
Suggesting:

> ```
>   echo "Pushing pre-release versions to Red Hat is not implemented (there is no unstable repository for them to live)"
> ```


<!-- Sent from Reviewable.io -->

rail reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/deploy-redhat/Dockerfile.in, line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MUu77-SBkjyYozSUEdk:-MUuYXRSC734RjIpvFiP:b31s187) ([raw file](https://github.com/cockroachdb/cockroach/blob/79cad1aeea145c3762ea28ae729395f0b876d88a/build/deploy-redhat/Dockerfile.in#L6)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

> ```
> LABEL summary="CockroachDB is a Distributed SQL database."
> LABEL description="CockroachDB is a PostgreSQL wire-compatable Distributed SQL database."
> ```
Were you able to ask docs if these were good strings to use?
</blockquote></details>

I asked in Slack, forgot to do this earlier :/
___
*[build/release/teamcity-publish-redhat-release.sh, line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MUu0XvX9TzJ7HeZeG-s:-MUuYpsy1_Ob1l2kwRCK:b-4o1upo) ([raw file](https://github.com/cockroachdb/cockroach/blob/79cad1aeea145c3762ea28ae729395f0b876d88a/build/release/teamcity-publish-redhat-release.sh#L10)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I don't think this does what the comment and message say this intends to do. Testing locally, this doesn't give the results expected here.
</blockquote></details>

Because I put the quotes in the wrong spot.

The current one looks better:
```
~
❯ ONE=v20.2.4-57-abcd345 TWO=v20.2.4-57 ; if [[ $ONE != "${TWO}-"* ]]; then echo Nope ; fi

~
❯ ONE=v20.2.4-57-abcd345 TWO=v20.2.4-5 ; if [[ $ONE != "${TWO}-"* ]]; then echo Nope ; fi
Nope

~
❯ ONE=v20.2.4-56-abcd345 TWO=v20.2.4-57 ; if [[ $ONE != "${TWO}-"* ]]; then echo Nope ; fi
Nope

~
❯ ONE=vvvvv20.2.4-57-abcd345 TWO=v20.2.4-57 ; if [[ $ONE != "${TWO}-"* ]]; then echo Nope ; fi
Nope
```
___
*[build/release/teamcity-publish-redhat-release.sh, line 15 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MUu2r61A7qJduPESmp9:-MUuZJuM2-GZAI0_bG6U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/79cad1aeea145c3762ea28ae729395f0b876d88a/build/release/teamcity-publish-redhat-release.sh#L15)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Suggesting:

> ```
>   echo "Pushing pre-release versions to Red Hat is not implemented (there is no unstable repository for them to live)"
> ```

</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jlinder reviewed (approved):
One final nit: could you align the PR description with the commit message? They are quite different.

Otherwise, once the feedback on the summary and description labels comes back from docs, this is good to go.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-:-MUykgnR5PR0CA2fFE2S:bomaugh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @rail)

___
*[build/release/teamcity-publish-redhat-release.sh, line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MUu0XvX9TzJ7HeZeG-s:-MUyi4tY9qyzKM5-fpf5:b-7mzt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/79cad1aeea145c3762ea28ae729395f0b876d88a/build/release/teamcity-publish-redhat-release.sh#L10)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Because I put the quotes in the wrong spot.

The current one looks better:
```
~
❯ ONE=v20.2.4-57-abcd345 TWO=v20.2.4-57 ; if [[ $ONE != "${TWO}-"* ]]; then echo Nope ; fi

~
❯ ONE=v20.2.4-57-abcd345 TWO=v20.2.4-5 ; if [[ $ONE != "${TWO}-"* ]]; then echo Nope ; fi
Nope

~
❯ ONE=v20.2.4-56-abcd345 TWO=v20.2.4-57 ; if [[ $ONE != "${TWO}-"* ]]; then echo Nope ; fi
Nope

~
❯ ONE=vvvvv20.2.4-57-abcd345 TWO=v20.2.4-57 ; if [[ $ONE != "${TWO}-"* ]]; then echo Nope ; fi
Nope
```
</blockquote></details>

:thumbsup: :tada: 
___
*[build/release/teamcity-publish-redhat-release.sh, line 15 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60007#-MUu2r61A7qJduPESmp9:-MUyi8irCBIiNLoCsNee:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/79cad1aeea145c3762ea28ae729395f0b876d88a/build/release/teamcity-publish-redhat-release.sh#L15)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->

rail commented:
bors r=jlinder

rail commented:
For some reason bors removed 2 lines from my commit message, what made the CI barf...

rail commented:
bors r=jlinder

jlinder commented:
> For some reason bors removed 2 lines from my commit message, what made the CI barf...

I think bors looks at the PR body (description) for the release justification message and not the individual commits.

rail commented:
> > For some reason bors removed 2 lines from my commit message, what made the CI barf...
> 
> I think bors looks at the PR body (description) for the release justification message and not the individual commits.

Ahhh, here we go. Now it makes more sense. Thanks for the explanation.
